EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -0,0 +1,252 @@
+  {
+    produces<mu2e::GenParticleCollection>();
+    produces<mu2e::PrimaryParticle>();
+    produces <MCTrajectoryCollection>();
+
+    if (_mean < 0) throw cet::exception("RANGE") << "CaloCalibGun.mean must be non-negative "<< '\n';
+  }
+
+  //================================================================
+  void CaloCalibGun::produce(art::Event& event) {
+    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
+    PrimaryParticle primaryParticles;
+    MCTrajectoryCollection mctc;
+
+    _cal =  &*GeomHandle<mu2e::DiskCalorimeter>();
+    //_nPipes          = _cal->caloInfo().getInt("nPipes");
+    _pipeRadius      = _cal->caloInfo().getDouble("pipeRadius");
+    _pipeTorRadius   = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _randomRad       = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _zPipeCenter     = _cal->disk(1).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(1).geomInfo().size().z()/2.0-_pipeRadius);//disk =1
+    // we normalize to the volume of the pipe (proportional to 2*pi*R if they have all the same radius) to draw a random number from which to generate the photons TODO - is this used?
+    double sumR(0);
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {sumR+=d; d = sumR; });
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {d /= sumR;});
+    
+    unsigned int nGen = _mean;
+    //Define the parameters of the pipes:
+    
+    // angle of large torus in degrees
+    std::vector<double> phi_lbd = _cal->caloInfo().getVDouble("LargeTorPhi");//{161.34, 149.50, 139.50, 132.07, 125.39}; 
+    // angle of small torus in degrees
+    std::vector<double> phi_sbd = _cal->caloInfo().getVDouble("smallTorPhi");//{84.63, 85.28, 85.79, 86.20, 86.53}; 
+    // angle of the end point
+    std::vector<double> phi_end = _cal->caloInfo().getVDouble("straitEndPhi");//{3.96, 10.53, 15.80, 20.16, 23.84};
+    // center position y of the small torus
+    std::vector<double> ysmall = _cal->caloInfo().getVDouble("yposition");//{432.2, 480.5, 524.3, 564.7, 602.5}; 
+    // radius of small torus
+    double radSmTor = _cal->caloInfo().getDouble("radSmTor");//41.0;
+    // first center position x of the small torus
+    double xsmall = _cal->caloInfo().getDouble("radSmTor");//71.0;
+    // distance of the small torus center
+    double xdistance = _cal->caloInfo().getDouble("xdistance");//60.0;
+    // inner radius of the manifold
+    double rInnerManifold = _cal->caloInfo().getDouble("rInnerManifold");//681.6; // 713.35 mm - 1.25 in (31.75 mm)
+    vector<float> sign{-1.0, 1.0};
+
+    for (unsigned int ig = 0; ig < nGen; ++ig) 
+    {
+      // only in the front of 2nd disk.
+      //int nDisk        = 1;
+      //_zPipeCenter     = _cal->disk(nDisk).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(nDisk).geomInfo().size().z()/2.0-_pipeRadius);
+
+      double xpipe, ypipe, zpipe;
+      //Pick position
```

Comment:
I believe you want to generate a RV that has values +-1. I naively recall that randFlat would return a value in [0;1]. If that's the case, then xsn will either be 0 or 1. Is that was it intended?

Reply:
yes, it is then it calls a vector of -1,+1 above with index 0 or 1